### PR TITLE
[Distributed] Initial distributed support, synced with latest main branch

### PR DESCRIPTION
**Pull request by ktoso**
https://github.com/apple/swift/pull/36302 but synced to main.

### CODE REVIEW

**Comment by ktoso** [comment]
@swift-ci build toolchain

**Comment by ktoso** [comment]
@swift-ci build toolchain

**Comment by ktoso** [comment]
@swift-ci build toolchain

**Comment by ktoso** [comment]
@swift-ci Please Build Toolchain macOS Platform

**Comment by ktoso** [comment]
@swift-ci Please Build Toolchain macOS Platform

**Comment by ktoso** [comment]
@swift-ci Please Build Toolchain macOS Platform

**Comment by ktoso** [comment]
@swift-ci please smoke test macOS

**Comment by ktoso** [comment]
@swift-ci please smoke test

File: stdlib/public/Concurrency/DistributedActor.swift [hunk 1]
```diff
@@ -0,0 +1,238 @@
...
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+import Swift
+@_implementationOnly import _SwiftConcurrencyShims
+
```
Comment (ktoso) [inline stdlib/public/Concurrency/DistributedActor.swift:15]: Do we need to we create a new module for these types? @DougGregor @rjmccall

**Comment by ktoso** [comment]
I _really_ would like to get to a place where we can merge this PR and keep it along but disabled the ongoing work rather than the current mode of operation...

Sadly after today's bringing this PR up to date with `main` we're again in some hell with seemingly unrelated issues... I'll work on this tomorrow but right now am at a bit of a loss about those:

`Assertion failed: (i->getTargetExecutor()->getType().is<BuiltinExecutorType>()), function visitHopToExecutorInst, file /Users/buildnode/jenkins/workspace/swift-PR-macos-smoke-test/branch-main/swift/lib/IRGen/IRGenSIL.cpp, line 6137.` triggering for:

```
00:27:51.295 ---
00:27:51.295 3.	While evaluating request IRGenRequest(IR Generation for module Swift)
00:27:51.295 4.	While emitting IR SIL function "@$ss7Float80V7exactlyABSgx_tcSzRzlufC".
00:27:51.295  for 'init(exactly:)' (at /Users/buildnode/jenkins/workspace/swift-PR-macos-smoke-test/branch-main/buildbot_incremental/swift-macosx-x86_64/stdlib/public/core/8/FloatingPointTypes.swift:4311:10)
00:27:51.295 Stack dump without symbol names (ensure you have llvm-symbolizer in your PATH or set the environment var `LLVM_SYMBOLIZER_PATH` to point to it):
00:27:51.295 0  swift-frontend           0x0000000106258c37 llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) + 39
00:27:51.295 1  swift-frontend           0x0000000106257ac8 llvm::sys::RunSignalHandlers() + 248
00:27:51.295 2  swift-frontend           0x0000000106259266 SignalHandler(int) + 262
00:27:51.295 3  libsystem_platform.dylib 0x00007fff6a4ac5fd _sigtramp + 29
00:27:51.295 4  libsystem_platform.dylib 0x00007ff18d63ee60 _sigtramp + 18446744014168860800
00:27:51.295 5  libsystem_c.dylib        0x00007fff6a37e808 abort + 120
00:27:51.295 6  libsystem_c.dylib        0x00007fff6a37dac6 err + 0
00:27:51.295 7  swift-frontend           0x00000001063d2db3 swift::SILInstructionVisitor<(anonymous namespace)::IRGenSILFunction, void>::visit(swift::SILInstruction*) (.cold.699) + 35
00:27:51.295 8  swift-frontend           0x0000000101c6e897 swift::SILInstructionVisitor<(anonymous namespace)::IRGenSILFunction, void>::visit(swift::SILInstruction*) + 77271
00:27:51.295 9  swift-frontend           0x0000000101c57151 (anonymous namespace)::IRGenSILFunction::emitSILFunction() + 7969
00:27:51.295 10 swift-frontend           0x0000000101c54c40 swift::irgen::IRGenModule::emitSILFunction(swift::SILFunction*) + 1696
00:27:51.295 11 swift-frontend           0x0000000101b3438b swift::irgen::IRGenerator::emitLazyDefinitions() + 1819
00:27:51.296 12 swift-frontend           0x0000000101c1ae47 swift::IRGenRequest::evaluate(swift::Evaluator&, swift::IRGenDescriptor) const + 2007
00:27:51.296 13 swift-frontend           0x0000000101c5411d swift::GeneratedModule swift::SimpleRequest<swift::IRGenRequest, swift::GeneratedModule (swift::IRGenDescriptor), (swift::RequestFlags)9>::callDerived<0ul>(swift::Evaluator&, std::__1::integer_sequence<unsigned long, 0ul>) const + 157
00:27:51.296 14 swift-frontend           0x0000000101c5403e swift::SimpleRequest<swift::IRGenRequest, swift::GeneratedModule (swift::IRGenDescriptor), (swift::RequestFlags)9>::evaluateRequest(swift::IRGenRequest const&, swift::Evaluator&) + 14
```

You added that assertion recently @rjmccall I think, would you have any guess / pointer perhaps?

If you have a spare moment @drexin perhaps you could try to have a look during your early day? I'll continue trying to figure this out as I get up early morning...

**Comment by DougGregor** [comment]
I think that means something is generating a `hop_to_executor` instruction that refers to an actor rather than its executor.

**Comment by drexin** [comment]
Hm, this is weird. This is failing in `Float80.init(exactly:)`. The particular block has the following SIL:

```
sil_scope 1 { loc "/Users/drexin/src/swift-source/build/Ninja-RelWithDebInfoAssert/swift-macosx-x86_64/stdlib/public/core/8/FloatingPointTypes.swift":4311:10 parent @$ss7Float80V7exactlyABSgx_tcSzRzlufC : $@convention(method) <τ_0_0 where τ_0_0 : BinaryInteger> (@in τ_0_0, @thin Float80.Type) -> Optional<Float80> }
sil_scope 2 { loc "/Users/drexin/src/swift-source/build/Ninja-RelWithDebInfoAssert/swift-macosx-x86_64/stdlib/public/core/8/FloatingPointTypes.swift":4312:5 parent 1 }
sil_scope 3 { loc "/Users/drexin/src/swift-source/build/Ninja-RelWithDebInfoAssert/swift-macosx-x86_64/stdlib/public/core/8/FloatingPointTypes.swift":4312:29 parent 2 }
sil_scope 4 { loc "/Users/drexin/src/swift-source/build/Ninja-RelWithDebInfoAssert/swift-macosx-x86_64/stdlib/public/core/8/FloatingPointTypes.swift":4332:14 parent 3 }
sil_scope 5 { loc "/Users/drexin/src/swift-source/build/Ninja-RelWithDebInfoAssert/swift-macosx-x86_64/stdlib/public/core/8/FloatingPointTypes.swift":4336:9 parent 4 }
bb29:                                             // Preds: bb28
  %162 = struct $Float80 (%158 : $Builtin.FPIEEE80) // users: %258, %259, %261, %173, %174, %176, %163
  hop_to_executor %162 : $Float80                 // id: %163
  %164 = builtin "assert_configuration"() : $Builtin.Int32 // users: %210, %166
  %165 = integer_literal $Builtin.Int32, 0        // user: %166
  %166 = builtin "cmp_eq_Int32"(%164 : $Builtin.Int32, %165 : $Builtin.Int32) : $Builtin.Int1 // user: %167
  cond_br %166, bb30, bb37                        // id: %167
```

Where is the `hop_to_executor` coming from?

**Comment by drexin** [comment]
This seems to be the code:

```swift
else {
  let extended = UInt(truncatingIfNeeded: value)
  _value = Builtin.uitofp_Int64_FPIEEE80(extended._value)
  guard self < 0x1.0p64 && UInt(self) == extended else {
    return nil
  }
}
```

File: lib/IDE/CodeCompletion.cpp [hunk 2]
```diff
@@ -4656,6 +4661,8 @@ class CompletionLookup final : public swift::VisibleDeclConsumer {
       return false;
     if (!IsConcurrencyEnabled && DeclAttribute::isConcurrencyOnly(DAK))
       return false;
+    if (!IsDistributedEnabled && DeclAttribute::isConcurrencyOnly(DAK))
```
Comment (drexin) [inline lib/IDE/CodeCompletion.cpp:4664]: should this be `isDistributedOnly`?

File: lib/Sema/TypeCheckConcurrency.cpp [hunk 3]
```diff
@@ -2628,6 +3109,10 @@ ActorIsolation ActorIsolationRequest::evaluate(
   if (auto nominal = value->getDeclContext()->getSelfNominalTypeDecl()) {
     if (nominal->isActor() &&
         (value->isInstanceMember() || isa<ConstructorDecl>(value))) {
+      defaultIsolation = nominal->isDistributedActor() ?
+                         ActorIsolation::forDistributedActorInstance(nominal) :
+                         ActorIsolation::forActorInstance(nominal);
+    } else if (isa<ConstructorDecl>(value)) {
       defaultIsolation = ActorIsolation::forActorInstance(nominal);
```
Comment (drexin) [inline lib/Sema/TypeCheckConcurrency.cpp:3116]: I think this is causing the error. It seems we are now treating every constructor as actor isolated, which would explain the `hop_to_executor` onto `Float80`.

```suggestion
```

Comment (ktoso) [reply lib/Sema/TypeCheckConcurrency.cpp:3116]: Thank you for digging in! let me check that

File: lib/IDE/CodeCompletion.cpp (hunk 2 above)
Comment (ktoso) [reply lib/IDE/CodeCompletion.cpp:4664]: oh thanks!

**Comment by ktoso** [comment]
Replaced by https://github.com/apple/swift/pull/37109

**Pull request closed without merging**